### PR TITLE
Added EmployeeManager.

### PR DESCRIPTION
**Pull request by TheStigger**
/CC @Fedorof, @ITsvetkoFF 

Implemented `EmployeeManagerDummy` for work with SSE data.
Dump with `dump.csv` placed in resources directory.
`db.csv` created after initialization of `EmployeeManagerDummy`. This file added to .gitignore to prevent conflicts between different development versions of file.

### CODE REVIEW

File: SSP/ssehelper/manager.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
+import csv
+import logging
+import os
+import shutil
+
+from os.path import join
+
+from django.conf import settings
+
+from SSP.ssp_project import settings as project_settings
```
Comment (Fedorof) [inline SSP/ssehelper/manager.py]: Just `from ssp_project import ...`. `SSP` prefix should not be used anymore. Rebase against master to get this fix

File: SSP/ssehelper/manager.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+from SSP.ssehelper.converters import list_to_employee, employee_to_list
+from SSP.ssehelper.employee import Employee, Workplace
+
+logger = logging.getLogger(__name__)
+
+
+class EmployeeIdError(ValueError):
+    """
+    add() function received Employee object with wrong id.
```
Comment (Fedorof) [inline SSP/ssehelper/manager.py]: Could you please change it to something like `An exception for incorrect Employee id`

File: SSP/ssehelper/manager.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+class Any:
+    """
+    Class which will return True after comparing with any other object.
+    """
+    def __eq__(self, other):
+        return True
+
+    def __str__(self):
+        return ''
```
Comment (Fedorof) [inline SSP/ssehelper/manager.py]: Why do you need it?

File: SSP/ssehelper/manager.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+    _workplace_cls = Workplace
+
+    def get_employees(self, employee_id=ANY, name=ANY, table=ANY,
+                      room=ANY, floor=ANY, project=ANY, vertical=ANY,
+                      position=ANY, location=ANY):
+        """
+        Check database for appropriate employees.
+        Returns list of appropriate _employee_cls objects.
+
+        :param employee_id: id of employee.
```
Comment (Fedorof) [inline SSP/ssehelper/manager.py]: Parameters descriptions are to plain. Either remove them or make more verbose

File: SSP/ssehelper/manager.py [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+        raise NotImplementedError("Not implemented yet.")
+
+
+class EmployeeManagerDummy(BaseEmployeeManager):
+    """
+    Implementation of AbstractEmployeeManager for Dummy data.
+    """
+
+    def __init__(self):
+        # TODO: we need to configure DUMP_PATH type in project settings.py
```
Comment (Fedorof) [inline SSP/ssehelper/manager.py]: Either deal with these TODOs or create userstories for them

File: SSP/ssehelper/manager.py [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+            join(settings.BASE_DIR, 'ssehelper', 'db.csv'))
+        if not os.path.isfile(self._file_path):
+            self._copy_dump()
+        self._pool = []
+        self._load_data()
+
+    def _copy_dump(self):
+        """
+        Copy file from DUMP_PATH to DUMMY_CSV_PATH.
+        New file used to store updated data.
```
Comment (Fedorof) [inline SSP/ssehelper/manager.py]: There should be a blank line between docstring title and body

File: SSP/ssehelper/manager.py [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+    def _employee_filter(self, employee_pattern):
+        """
+        Filter _pool for appropriate _employee_cls objects.
+        Return list of objects.
+
+        :return: list of appropriate _employee_cls objects.
+        :rtype: list.
+        """
+        return list(filter(lambda employee: employee == employee_pattern,
+                           self._pool))
```
Comment (Fedorof) [inline SSP/ssehelper/manager.py]: It goes better:

    return [e for e in self._pool if e == employee_pattern]

File: SSP/ssehelper/tests/resources/dump.csv [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+273;Global Talents Acquisition;Global Talent Acquisition\Global Talents Acquisition\DevCenter's Operations\Kyiv;Senior Recruiter;Florentino Wagoner;7;123;Kyiv1
+274;Global Talents Acquisition;Global Talent Acquisition\Global Talents Acquisition;Senior Recruiter;Barabara Trimble;1;623-kubrik;Kyiv1
+275;Enterprise Technology B;Operations\Delivery Verticals\Enterprise Technology Vertical\Enterprise Technology B\xReserve\Internship\Ginger Ale\GD SDK Kyiv, GD Server;Software Engineer Trainee;Khalilah Bravo;9;328;Kyiv1
+276;Healthcare B;Operations\Delivery Verticals\Healthcare Vertical\Healthcare B\noRules;Software Engineer Trainee;Antonia Nunley;1;125;Kyiv1
+277;Global Talents Acquisition;Global Talent Acquisition\Global Talents Acquisition\DevCenter's Operations\Kyiv;Talent Acquisition Manager;Theola Starks;6;123;Kyiv1
+278;SoftServe IT Academy;Human Resources\HRD&SSU\SoftServe  University\SoftServe IT Academy;Teacher;Lin Simonson;1;111;Kyiv1
+279;Language School;Human Resources\HRD&SSU\SoftServe  University\Language School;Teacher;Argentina Knutson;1;327;Kyiv1
+280;Language School;Human Resources\HRD&SSU\SoftServe  University\Language School;Teacher;Chris Salisbury;1;426;Kyiv1
```
Comment (Fedorof) [inline SSP/ssehelper/tests/resources/dump.csv]: Is not it equal to the file above? Why don't use both? Or make this one smaller (only 3 lines)

File: SSP/ssehelper/tests/test_manager_dummy.py [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+from os.path import join
+
+from django.test import override_settings, TestCase
+from django.conf import settings
+
+from SSP.ssp_project import settings as project_settings
+from SSP.ssehelper.manager import get_manager, EmployeeManagerDummy, \
+    EmployeeIdError
+
+settings.configure(project_settings)
```
Comment (Fedorof) [inline SSP/ssehelper/tests/test_manager_dummy.py]: It should not be there.

File: SSP/ssehelper/tests/test_manager_dummy.py [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+
+
+@override_settings(
+    DUMMY_CSV_PATH=join(settings.BASE_DIR, 'ssehelper', 'tests', 'resources',
```
Comment (Fedorof) [inline SSP/ssehelper/tests/test_manager_dummy.py]: Is it cleaned up after tests? And maybe it's better to use `NamedTemporaryFile`.

File: SSP/ssehelper/tests/test_manager_dummy.py [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+    # TODO: add tests for employees without workplace
+
+    def setUp(self):
+        """
+        Create EmployeeManagerDummy instance.
+        Create object of employee with id = 1.
+        """
+        self.manager = get_manager()
+
+        self.workplace_of_employee_with_id_1 = self.manager._workplace_cls(
```
Comment (Fedorof) [inline SSP/ssehelper/tests/test_manager_dummy.py]: It's not a good idea to call private methods. Use classes directly. And `workplace_of_employee_with_id_1` is a bad name

File: SSP/ssehelper/tests/test_manager_dummy.py [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+            position='Abiliton Application Architect',
+            project='IT\\TRSY\\TRSY Management\\Software Architecture',
+            vertical='TRSY',
+            workplaces=[self.workplace_of_employee_with_id_1])
+
+    def test_manager_type(self):
+        """
+        Check type of self.manager. It should be EmployeeManagerDummy.
+        """
+        self.assertEqual(True, type(self.manager) is EmployeeManagerDummy)
```
Comment (Fedorof) [inline SSP/ssehelper/tests/test_manager_dummy.py]: Type checks is a bad practice. Especially for abstract factory.

File: SSP/ssehelper/tests/test_manager_dummy.py [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+        Test automatic creation of test_db.csv.
+        """
+        self.assertEqual(True, os.path.isfile(settings.DUMMY_CSV_PATH))
+
+    def test_db_copying(self):
+        """
+        Test copying dump.csv file content to test_db.csv.
+        """
+        self.assertEqual(True, filecmp.cmp(settings.DUMMY_CSV_PATH,
+                                           settings.DUMP_PATH))
```
Comment (Fedorof) [inline SSP/ssehelper/tests/test_manager_dummy.py]: Stop testing obvious things, focus on behaviour.

File: SSP/ssehelper/tests/test_manager_dummy.py [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+
+    def test_get_employees_by_id(self):
+        """
+        Test get_employees() method by employee object pattern with
+        specified id.
+        """
+        employees = self.manager.get_employees(employee_id='1')
+        self.assertEqual(self.employee_with_id_1, employees[0])
```
Comment (Fedorof) [inline SSP/ssehelper/tests/test_manager_dummy.py]: Perhaps `self.assertEqual([self.employee_with_id_1], employees)`

File: SSP/ssehelper/tests/test_manager_dummy.py [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+
+    def test_get_employees_by_absent_id(self):
+        """
+        Test get_employees() method by employee object pattern with
+        specified id that is absent in database.
+        """
+        self.assertEqual(0, len(self.manager.get_employees(
+            employee_id='0')))
```
Comment (Fedorof) [inline SSP/ssehelper/tests/test_manager_dummy.py]: self.assertEqual(self.manager.get_employees(employee_id=0), [])

File: SSP/ssehelper/tests/test_manager_dummy.py [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+
+    def test_get_employees_by_project(self):
+        """
+        Test get_employees() method by employee object pattern with
+        specified project.
+        """
+        get_result = self.manager.get_employees(
+            project='IT\\TRSY\\TRSY Management\\Software Architecture')
+        self.assertEqual(True, self.employee_with_id_1 == get_result[0])
```
Comment (Fedorof) [inline SSP/ssehelper/tests/test_manager_dummy.py]: Stop using `True`, `False`, `None` in `assertEqual`. Just check the items directly

File: SSP/ssehelper/tests/test_manager_dummy.py [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+
+    def test_get_employees_by_parent_project(self):
+        """
+        Test get_employees() method by employee object pattern
+        with specified parent id.
+        """
+        get_result = self.manager.get_employees(project='IT\\')
+        self.assertEqual(True, get_result[0] == self.employee_with_id_1 and
+                         get_result[1].id_sse == '92' and
+                         get_result[2].id_sse == '255')
```
Comment (Fedorof) [inline SSP/ssehelper/tests/test_manager_dummy.py]: Just use 3 different `assertEqual`

File: SSP/ssehelper/tests/test_manager_dummy.py [hunk 18]
```diff
@@ -0,0 +1,232 @@
...
+    def test_get_employees_with_multiple_workplaces(self):
+        """
+        Test get_employees() method by employee object pattern
+        and check result employee with multiple workplaces.
+        """
+        result = self.manager.get_employees(employee_id='44')[0]
+        self.assertEqual(2, len(result.workplaces),
+                         'Incorrect number of workplaces.')
+        self.assertNotEqual(result.workplaces[0], result.workplaces[1],
+                            'Workplaces can\'t be equal.')
```
Comment (Fedorof) [inline SSP/ssehelper/tests/test_manager_dummy.py]: "Workplaces are unexpectedly equal: {!r} == {!r}".format(*result.workplaces)

File: SSP/ssehelper/tests/test_manager_dummy.py [hunk 19]
```diff
@@ -0,0 +1,232 @@
...
+        """
+        Test changes of id after execution of add().
+        """
+        new_employee = self.employee_with_id_1
+        new_employee.name = 'New employee'
+        new_employee.id_sse = None
+
+        self.manager.add(new_employee)
+
+        self.assertNotEqual(None, new_employee.id_sse)
```
Comment (Fedorof) [inline SSP/ssehelper/tests/test_manager_dummy.py]: `assertIsNotNone`

File: SSP/ssehelper/tests/test_manager_dummy.py [hunk 20]
```diff
@@ -0,0 +1,232 @@
...
+        Test raising exceptions for non-numerical id.
+        """
+        new_employee = self.employee_with_id_1
+        new_employee.name = 'New employee'
+        new_employee.id_sse = 'abc'
+
+        with self.assertRaises(ValueError):
+            self.manager.add(new_employee)
+
+    def tearDown(self):
```
Comment (Fedorof) [inline SSP/ssehelper/tests/test_manager_dummy.py]: Place it right after `setUp`, please

File: SSP/ssehelper/manager.py (hunk 3 above)
Comment (TheStigger) [inline SSP/ssehelper/manager.py]: @Fedorof do you mean __str__() method or whole class?

File: SSP/ssehelper/tests/test_manager_dummy.py (hunk 9 above)
Comment (TheStigger) [inline SSP/ssehelper/tests/test_manager_dummy.py]: @Fedorof but how can I configure django settings?

File: SSP/ssehelper/tests/test_manager_dummy.py (hunk 11 above)
Comment (TheStigger) [inline SSP/ssehelper/tests/test_manager_dummy.py]: @Fedorof but EmployeeManager uses private classes. Is it a good idea to create objects from different class references?

File: SSP/ssehelper/manager.py (hunk 3 above)
Comment (Fedorof) [inline SSP/ssehelper/manager.py]: `__str__`

File: SSP/ssehelper/tests/test_manager_dummy.py (hunk 9 above)
Comment (Fedorof) [inline SSP/ssehelper/tests/test_manager_dummy.py]: Don't `django.conf.settings` already contain the project settings? If not, there are problems with tests or the project configuration.

File: SSP/ssehelper/tests/test_manager_dummy.py (hunk 11 above)
Comment (Fedorof) [inline SSP/ssehelper/tests/test_manager_dummy.py]: `Employee` and `Workplace` are not private.

File: SSP/ssehelper/manager.py [hunk 21]
```diff
@@ -0,0 +1,265 @@
...
+        """
+        Check new employee id. If id is None, then assign new id.
+        If id already specified, then it should be unique.
+
+        :param new_employee: object of new employee
+        :type new_employee: _employee_cls.
+        :raise EmployeeIdError, ValueError: raises error in case of wrong id.
+        """
+        if new_employee.id_sse is None:
+            new_employee.id_sse = str(
```
Comment (Fedorof) [inline SSP/ssehelper/manager.py]: 1. Use `int` for numeric values.
2. At least use this:

    new_employee.id_sse = str(max(int(e.id_sse) for e in self._pool) + 1)

File: SSP/ssehelper/manager.py (hunk 3 above)
Comment (TheStigger) [inline SSP/ssehelper/manager.py]: We check project by this condition: `self.project.startswith(str(other.project))`. For example we have full project path **A/B/C**. And when we search for **A/B** project we expect to get all employee with project **A/B** and all its successors. In this case `startswith()` method uses `__str__`. I don't see better approach to solve this problem.

File: SSP/ssehelper/manager.py (hunk 4 above)
Comment (TheStigger) [inline SSP/ssehelper/manager.py]: @Fedorof can I leave description for some not obvious parameters and remove rest?

Comment (Fedorof) [inline SSP/ssehelper/manager.py]: Sure. Just notice that `employee_id is an id of employee` is a bad description

File: SSP/ssehelper/manager.py (hunk 3 above)
Comment (Fedorof) [inline SSP/ssehelper/manager.py]: This approach should be changed. Could you please point me at that piece of code? And for now at least leave a verbose comment to `__str__` to explain why and *where* it is used

File: SSP/ssehelper/tests/test_manager_dummy.py (hunk 11 above)
Comment (TheStigger) [inline SSP/ssehelper/tests/test_manager_dummy.py]: @Fedorof
> And `workplace_of_employee_with_id_1` is a bad name.

What would be better? Maybe I need create workplace directly while creating employee?

Comment (Fedorof) [inline SSP/ssehelper/tests/test_manager_dummy.py]: Just having `self.workplace`and `self.employee` in `setUp` should be Ok

**Pull request merged**